packages/relay-compiler/transforms/RelayGenerateIDFieldTransform.js

```diff
@@ -11,104 +11,113 @@
-  hasID,
   implementsInterface,
   isAbstractType,
   mayImplement,
 } = SchemaUtils;
 
 const ID = 'id';
+const ID_KEY = '__id';
 const ID_TYPE = 'ID';
 const NODE_TYPE = 'Node';
 
-type State = {
-  idField: ScalarField,
-};
-
 /**
- * A transform that adds an `id` field on any type that has an id field but
- * where there is no unaliased `id` selection.
+ * A transform that adds a `__id` field on any type that has a `Node` or `id`
+ * field but where there is no unaliased `__id` selection.
  */
 function relayGenerateIDFieldTransform(
   context: CompilerContext,
 ): CompilerContext {
-  const idType = assertLeafType(context.serverSchema.getType(ID_TYPE));
-  const idField: ScalarField = {
-    kind: 'ScalarField',
-    alias: (null: ?string),
-    args: [],
-    directives: [],
-    handles: null,
-    metadata: null,
-    name: ID,
-    type: idType,
-  };
-  const state = {
-    idField,
-  };
-  return IRTransformer.transform(
-    context,
-    {
-      LinkedField: visitLinkedField,
-    },
-    () => state,
-  );
+  return IRTransformer.transform(context, {
+    LinkedField: visitNodeWithSelections,
+    Fragment: visitNodeWithSelections,
+  });
 }
 
-function visitLinkedField(field: LinkedField, state: State): LinkedField {
-  const transformedNode = this.traverse(field, state);
+function visitNodeWithSelections<T: Fragment | LinkedField>(node: T): T {
+  const transformedNode = this.traverse(node);
 
-  // If the field already has an unaliased `id` field, do nothing
-  if (hasUnaliasedSelection(field, ID)) {
+  // If the field already has an `__id` selection, do nothing.
+  if (hasSelection(node, '__id')) {
     return transformedNode;
   }
 
   const context = this.getContext();
   const schema = context.serverSchema;
-  const unmodifiedType = assertCompositeType(getRawType(field.type));
+  const unmodifiedType = assertCompositeType(getRawType(node.type));
+  const idFieldDefinition = getIDFieldDefinition(schema, unmodifiedType);
 
-  // If the field type has an `id` subfield add an `id` selection
-  if (canHaveSelections(unmodifiedType) && hasID(schema, unmodifiedType)) {
+  // If the field type has a ID field add a selection for that field.
+  if (idFieldDefinition && canHaveSelections(unmodifiedType)) {
     return {
       ...transformedNode,
-      selections: [...transformedNode.selections, state.idField],
+      selections: [
+        ...transformedNode.selections,
+        buildSelectionFromFieldDefinition(idFieldDefinition),
+      ],
     };
   }
 
-  // If the field type is abstract, then generate a `... on Node { id }`
-  // fragment if *any* concrete type implements Node. Then generate a
-  // `... on PossibleType { id }` for every concrete type that does *not*
-  // implement `Node`
+  // - If the field type is abstract, then generate a `... on Node { __id: id }`
+  //   fragment if *any* concrete type implements `Node`. Then generate a
+  //   `... on PossibleType { __id: id }` for every concrete type that does
```

Comment:
what if `PossibleType` doesn't have an id field? can we add test coverage for this?